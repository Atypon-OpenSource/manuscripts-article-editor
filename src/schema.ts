export const manuscripts = {
  title: 'manuscripts schema',
  version: 0,
  description: 'describes a manuscript',
  type: 'object',
  properties: {
    title: {
      type: 'string',
    },
  },
  required: ['title'],
}

export const sections = {
  title: 'sections schema',
  version: 1,
  description: 'describes a section',
  type: 'object',
  properties: {
    title: {
      type: 'string',
    },
    content: {
      type: 'string',
    },
    manuscript: {
      type: 'string',
    },
  },
  required: ['title'],
}

export const people = {
  title: 'people schema',
  version: 0,
  description: 'describes a person',
  type: 'object',
  properties: {
    name: {
      type: 'string',
    },
    surname: {
      type: 'string',
    },
  },
  required: ['name'],
}

export const groups = {
  title: 'groups schema',
  version: 0,
  description: 'describes a group',
  type: 'object',
  properties: {
    name: {
      type: 'string',
    },
    description: {
      type: 'string',
    },
  },
  required: ['name'],
}
