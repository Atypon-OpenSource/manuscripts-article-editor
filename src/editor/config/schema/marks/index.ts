import { MarkSpec } from 'prosemirror-model'
import { marks } from 'prosemirror-schema-basic'
import { StringMap } from '../../types'

const subscript: MarkSpec = {
  group: 'position',
  excludes: 'superscript',
  parseDOM: [{ tag: 'sub' }, { style: 'vertical-align=sub' }],
  toDOM: () => ['sub'],
}

const superscript: MarkSpec = {
  group: 'position',
  excludes: 'subscript',
  parseDOM: [{ tag: 'sup' }, { style: 'vertical-align=super' }],
  toDOM: () => ['sup'],
}

const strikethrough: MarkSpec = {
  parseDOM: [
    { tag: 'strike' },
    { style: 'text-decoration:line-through' },
    { style: 'text-decoration-line:line-through' },
  ],
  toDOM: () => [
    'span',
    {
      style: 'text-decoration-line:line-through',
    },
  ],
}

const underline: MarkSpec = {
  parseDOM: [{ tag: 'u' }, { style: 'text-decoration:underline' }],
  toDOM: () => [
    'span',
    {
      style: 'text-decoration:underline',
    },
  ],
}

const combinedMarks: StringMap<MarkSpec> = {
  ...marks,
  subscript,
  superscript,
  strikethrough,
  underline,
}

export default combinedMarks
