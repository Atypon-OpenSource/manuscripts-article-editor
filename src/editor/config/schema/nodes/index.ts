import { NodeSpec } from 'prosemirror-model'
import { nodes as basic } from 'prosemirror-schema-basic'
import { bibliographyElement } from './bibliography_element'
import { bibliographySection } from './bibliography_section'
import { caption } from './caption'
import { citation } from './citation'
import { crossReference } from './cross_reference'
import { doc } from './doc'
import { equation } from './equation'
import { equationElement } from './equation_element'
import { figcaption } from './figcaption'
import { figureElement } from './figure_element'
import { footnote } from './footnote'
import { footnotesElement } from './footnotes_element'
import { inlineEquation } from './inline_equation'
import { inlineFootnote } from './inline_footnote'
import { ListNodes, listNodes } from './list'
import { listing } from './listing'
import { listingElement } from './listing_element'
import { manuscript } from './manuscript'
import { paragraph } from './paragraph'
import { section } from './section'
import { sectionTitle } from './section_title'
import { TableNodes, tableNodes } from './table'
import { tableElement } from './table_element'
import { title } from './title'
import { tocElement } from './toc_element'
import { tocSection } from './toc_section'

export type Nodes =
  | 'bibliography_element'
  | 'bibliography_section'
  | 'blockquote'
  | 'caption'
  | 'citation'
  | 'cross_reference'
  | 'doc'
  | 'equation'
  | 'equation_element'
  | 'figcaption'
  // | 'figure' TODO
  | 'figure_element'
  | 'footnote'
  | 'footnotes_element'
  | 'hard_break'
  | 'horizontal_rule'
  | 'inline_equation'
  | 'inline_footnote'
  | 'listing'
  | 'listing_element'
  | 'manuscript'
  | 'paragraph'
  | 'section'
  | 'section_title'
  | 'section_title'
  | 'table_element'
  | 'text'
  | 'title'
  | 'toc_element'
  | 'toc_section'
  | ListNodes
  | TableNodes

const combinedNodes: { [key in Nodes]: NodeSpec } = {
  ...listNodes,
  ...tableNodes,
  manuscript,
  bibliography_element: bibliographyElement,
  bibliography_section: bibliographySection,
  blockquote: basic.blockquote,
  caption,
  citation,
  cross_reference: crossReference,
  doc,
  equation,
  equation_element: equationElement,
  figcaption,
  figure_element: figureElement,
  footnote,
  footnotes_element: footnotesElement,
  hard_break: basic.hard_break,
  horizontal_rule: basic.horizontal_rule,
  inline_equation: inlineEquation,
  inline_footnote: inlineFootnote,
  listing,
  listing_element: listingElement,
  paragraph,
  section,
  section_title: sectionTitle,
  table_element: tableElement,
  text: basic.text,
  title,
  toc_element: tocElement,
  toc_section: tocSection,
}

export default combinedNodes
