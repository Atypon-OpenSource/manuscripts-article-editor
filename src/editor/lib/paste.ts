import { Slice } from 'prosemirror-model'

const removeFirstParagraphIfEmpty = (slice: Slice) => {
  const firstChild = slice.content.firstChild

  if (
    firstChild &&
    firstChild.type.name === 'paragraph' &&
    firstChild.textContent === ''
  ) {
    slice.content = slice.content.cut(firstChild.nodeSize)
  }
}

// remove `id` from pasted content
const removeIDs = (slice: Slice) => {
  slice.content.descendants(node => {
    if (node.attrs.id) {
      node.attrs.id = null
    }
  })
}

export const transformPasted = (slice: Slice): Slice => {
  removeFirstParagraphIfEmpty(slice)

  removeIDs(slice)

  return slice
}
