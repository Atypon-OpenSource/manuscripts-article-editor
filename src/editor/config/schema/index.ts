import { Schema } from 'prosemirror-model'

import marks, { Marks } from './marks'
import nodes, { Nodes } from './nodes'

export default new Schema<Nodes, Marks>({ nodes, marks })
