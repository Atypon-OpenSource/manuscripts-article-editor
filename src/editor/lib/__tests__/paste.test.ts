import { Slice } from 'prosemirror-model'
import schema from '../../config/schema'
import { transformPasted } from '../paste'

test('transformPasted handler', () => {
  const slice = Slice.fromJSON(schema, {
    content: [
      {
        type: 'paragraph',
        attrs: {},
        content: [{ type: 'hard_break' }],
      },
      {
        type: 'paragraph',
        attrs: {},
        content: [
          {
            type: 'text',
            text:
              'These rhythmic patterns then sum together to create the signals that muscles need to carry out the movements.',
          },
        ],
      },
    ],
    openStart: 1,
    openEnd: 1,
  })

  expect(slice.content.childCount).toBe(2)

  const result = transformPasted(slice)

  expect(result.content.childCount).toBe(1)
  expect(result.content.size).toBe(111)

  expect(result).toMatchSnapshot('transform-pasted')
})
