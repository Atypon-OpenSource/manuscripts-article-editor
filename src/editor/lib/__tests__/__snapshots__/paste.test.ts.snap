// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformPasted handler: transform-pasted 1`] = `
Object {
  "content": Array [
    Object {
      "attrs": Object {
        "id": null,
        "paragraphStyle": "",
        "placeholder": "",
      },
      "content": Array [
        Object {
          "text": "These rhythmic patterns then sum together to create the signals that muscles need to carry out the movements.",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
  "openEnd": 1,
  "openStart": 1,
}
`;
